alicloud/resource_alicloud_ssl_vpn_client_cert.go

```diff
@@ -0,0 +1,149 @@
+package alicloud
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/vpc"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAliyunSslVpnClientCert() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunSslVpnClientCertCreate,
+		Read:   resourceAliyunSslVpnClientCertRead,
+		Update: resourceAliyunSslVpnClientCertUpdate,
+		Delete: resourceAliyunSslVpnClientCertDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"ssl_vpn_server_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateInstanceName,
+			},
+
+			"status": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAliyunSslVpnClientCertCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	var sslVpnClientCert *vpc.CreateSslVpnClientCertResponse
+
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
+		args := buildAliyunSslVpnClientCertArgs(d, meta)
+
+		resp, err := client.vpcconn.CreateSslVpnClientCert(args)
+		if err != nil {
+			if IsExceptedError(err, VpnConfiguring) {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		sslVpnClientCert = resp
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("Create ssl vpn client cert got an error :%#v", err)
+	}
+
+	d.SetId(sslVpnClientCert.SslVpnClientCertId)
+
+	err = client.WaitForSslVpnClientCert(d.Id(), Ssl_Cert_Normal, 60)
+	if err != nil {
+		return fmt.Errorf("Timeout when WaitForSslVpnClientCertAvailable")
+	}
+
+	return resourceAliyunSslVpnClientCertRead(d, meta)
+}
+
+func resourceAliyunSslVpnClientCertRead(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*AliyunClient)
+
+	resp, err := client.DescribeSslVpnClientCert(d.Id())
+
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("status", resp.Status)
+
+	return nil
+}
+
+func resourceAliyunSslVpnClientCertUpdate(d *schema.ResourceData, meta interface{}) error {
+	attributeUpdate := false
+	request := vpc.CreateModifySslVpnClientCertRequest()
+	request.SslVpnClientCertId = d.Id()
+
+	if d.HasChange("name") {
+		request.Name = d.Get("name").(string)
+		attributeUpdate = true
+	}
+
+	if attributeUpdate {
```

Comment:
I think attributeUpdate is redundant and line 96, 97 and 105-107 can be moved into line 99 block.